sanic/blueprints.py

```diff
@@ -148,17 +149,33 @@ def signal(self, event: str, *args, **kwargs):
     def copy(
         self,
         name: str,
-        url_prefix: str = "",
-        version: Optional[Union[int, str, float]] = None,
-        strict_slashes: Optional[bool] = None,
-        version_prefix: str = "/v",
+        url_prefix: str = _default,
+        version: Optional[Union[int, str, float]] = _default,
+        version_prefix: str = _default,
+        strict_slashes: Optional[bool] = _default,
     ):
+        """
+        Copy a blueprint instance with some optional parameters to 
+        override the values of attributes in the old instance.
+
+        :param name: unique name of the blueprint
+        :param url_prefix: URL to be prefixed before all route URLs
+        :param host: IP Address of FQDN for the sanic server to use.
+        :param version: Blueprint Version
+        :param version_prefix: the prefix of the version number shown in the URL.
+        :param strict_slashes: Enforce the API urls are requested with a
+            training */*
+        """
         new_bp = deepcopy(self)
```

Comment:
I am not sure `deepcopy` alone will do the job. Once a blueprint is registered to an app, you might have problems with this. Perhaps the solution might just be to say that once it is registered there can be no copying. In which case we would need to make sure that is the case and raise an exception up front. (check bp._apps).

Reply:
Thank you @ahopkins.
Let me try to find the reason and potential solution for this problem.

Reply:
@ahopkins I found a workaround to remove all registered info in the the new bp instance (see code below for details). That way we can deepcopy the instance after it is being registered with some Sanic app. What do you think?
```
temp_attrs = {
    "_apps": self._apps,
    "routes": self.routes,
    "websocket_routes": self.websocket_routes,
   "middlewares": self.middlewares,
    "exceptions": self.exceptions,
    "listeners": self.listeners,
}
self._apps = set()
self.routes = []
self.websocket_routes = []
self.middlewares = []
self.exceptions = []
self.listeners = {}

new_bp = deepcopy(self)
        
#...... more code here ......

for key, value in temp_attrs.items():
    setattr(self, key, value)
```

Reply:
That works for me. It might be nice to make it a little less verbose:

```python
set_aside = (
    ("_apps", set),
    ("routes", list),
    ...
)

temp_attrs = {}

for attr, factory in set_aside:
    temp_attrs[attr] = getattr(self, attr)
    setattr(self, attr, factory())

...
```

Reply:
The more I think... I like this idea.

```python
def __init__(...):
    # set variables from args, remove the rest of the code that is now in reset
    self.reset()

def copy(self, ..., with_ctx=False):
    new_bp = deepcopy(self)
    new_bp.reset()

    options = OVERRIDE_VALUES_HERE

    for app in self._apps:
        new_bp.register(app, options)

def reset(self):
    # remove all of these from __init__
    self._apps: Set[Sanic] = set()
    self.exceptions: List[RouteHandler] = []
    self.listeners: Dict[str, List[ListenerType]] = {}
    self.middlewares: List[MiddlewareType] = []
    self.routes: List[Route] = []
    self.statics: List[RouteHandler] = []
    self.websocket_routes: List[Route] = []
```

Reply:
@ahopkins I think `reset` is a good idea, however, we will still have the problem with `deepcopy` after the old bp being registered... We cannot directly reset old bp before copying it because we don't want to change anything in the old bp, so, we probably still need something like the temporarily reset of registration info mentioned above. We can have an option to allow users to keep the registration info from the old bp by doing something like:

```
if keep_old_registerations:
    for app in temp_attrs["_apps"]:
        app.blueprint(new_bp, **options)
```

Reply:
By the way, I didn't understand what's meaning by `with_ctx` and `self.statics`.

I think `self.statics` is related to registration. Should we reset it during copying?

Reply:
> we probably still need something like the temporarily reset of registration info mentioned above. We can backup the attrs, call `reset`, and then restore attrs after copying. We can also have an option to allow users to keep the registration info from the old bp by doing something like:

Perhaps we are overthinking that. It is only a handful of fields:

```python
def copy(self, name, ..., with_ctx=False, with_registration=True):
    options = {
        "host": host,
        "strict_slashes": strict_slashes,
        "url_prefix": url_prefix,
        "version": version,
        "version_prefix": version_prefix,
    }

    new_bp = Blueprint(name)
    for attr, value in options.items():
        value = getattr(self, attr) if value is _default else value
        setattr(attr, new_bp, value)

    if with_registration:
        for app in self._apps:
            new_bp.register(app, options)

    if not with_ctx:
        new_bp = SimpleNamespace()
```



> I think sometimes users don't want to keep registration from the old bp, so this could be an optional feature.

I suppose. Although, it is probably not often that someone is building more than one app in a given impl. More often than not, there is only one application to even register to. So, I would default it to `True`.



> I didn't understand what's meaning by `with_ctx` and `self.statics`

`with_ctx` meaning that the `blueprint.ctx` should be copied. Default to `False`

`self.statics` are static file handlers: `bp.static(...)`

Reply:
@ahopkins I updated the branch, I consider we solved this issue. Can you review it and see if we can resolve this? Thanks! Please let me know if you have any suggestion.